src/Featurizers/TruncatedSVDFeaturizer.h

```diff
@@ -0,0 +1,631 @@
+    Scalar v2 = static_cast<Scalar>(std::rand() + static_cast<Scalar>(1)) / (static_cast<Scalar>(RAND_MAX+static_cast<Scalar>(2)));
+    Scalar len = sqrt(static_cast<Scalar>(-2) * log(v1));
+    x = len * cos(static_cast<Scalar>(2) * PI * v2);
+    y = len * sin(static_cast<Scalar>(2) * PI * v2);
+}
+	
+template<typename MatrixType>
+inline void sample_gaussian(MatrixType& mat) {
+
+    typedef typename MatrixType::Index Index;
+		
+    for(Index i = 0; i < mat.rows(); ++i)
+    {
+        for(Index j = 0; j+1 < mat.cols(); j += 2)
+            sample_gaussian(mat(i, j), mat(i, j+1));
+        if(mat.cols() % 2)
+            sample_gaussian(mat(i, mat.cols()-1), mat(i, mat.cols()-1));
+    }
+}
+	
+template<typename MatrixType>
+inline void gram_schmidt(MatrixType& mat) {
+	
+    typedef typename MatrixType::Scalar Scalar;
+    typedef typename MatrixType::Index Index;
+
+#if (defined __clang__)
+#   pragma clang diagnostic push
+#   pragma clang diagnostic ignored "-Wdouble-promotion"
+#endif
+		
+    static const Scalar EPS(1E-4f);
+
+#if (defined __clang__)
+#   pragma clang diagnostic pop
+#endif
+		
+    for(Index i = 0; i < mat.cols(); ++i) {
+        for(Index j = 0; j < i; ++j) {
+            Scalar r = mat.col(i).dot(mat.col(j));
+            mat.col(i) -= r * mat.col(j);
+        }
+			
+        Scalar norm = mat.col(i).norm();
+			
+        if(norm < EPS) {
+            for(Index k = i; k < mat.cols(); ++k)
+                mat.col(k).setZero();
+            return;
+        }
+        mat.col(i) /= norm;
+    }
+}
+
+} //anonymous namespace
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         SVDComponentsAnnotationData
+///  \brief         Contains SVD components: Eigenvalues and Eigenvectors
+template <typename T>
+class SVDComponentsAnnotationData {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Data
+    // |
+    // ----------------------------------------------------------------------
+
+    T const                                          SingularValues;
+    T const                                          SingularVectors;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    SVDComponentsAnnotationData(T singularvalues, T singularvectors);
+    ~SVDComponentsAnnotationData(void) = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(SVDComponentsAnnotationData);
+};
+
+namespace ComponentsDetails {
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         SVDTrainingOnlyPolicy
+///  \brief         `SVDComponentsEstimator` implementation details.
+///
+template <typename InputT, typename TransformedT>
```

Comment:
If InputT and TransformedT are the same, let's make them a single template type.